src/qobj/quantum_object_base.jl

```diff
@@ -0,0 +1,216 @@
+#=  
+This file defines the AbstractQuantumObject structure, all the type structures for AbstractQuantumObject, and fundamental functions in Julia standard library:  
+    - Base: show, length, size, eltype, getindex, setindex!, isequal, :(==), isapprox  
+=#
+
+export AbstractQuantumObject
+export QuantumObjectType,
+    BraQuantumObject,
+    KetQuantumObject,
+    OperatorQuantumObject,
+    OperatorBraQuantumObject,
+    OperatorKetQuantumObject,
+    SuperOperatorQuantumObject
+export Bra, Ket, Operator, OperatorBra, OperatorKet, SuperOperator
+
+@doc raw"""
+    abstract type AbstractQuantumObject{DataType,ObjType,N}
+
+Abstract type for all quantum objects like [`QuantumObject`](@ref) and [`QuantumObjectEvolution`](@ref).
+
+# Example
+```
+julia> sigmax() isa AbstractQuantumObject
+true
+```
+"""
+abstract type AbstractQuantumObject{DataType,ObjType,N} end
```

Comment:
We should remove the parameter `N`, namely
```julia
abstract type AbstractQuantumObject{DataType,ObjType} end
```

Reply:
Or maybe we can add N,M. Meaning the input and output dimension. But maybe it is better to discuss about it after this PR